feat(reports): purchase price report

jniles opened the pull request:
This report shows the historical purchase price of an item since there is no easy way to get this information in BHIMA currently.

Close #5519.

![image](https://user-images.githubusercontent.com/896472/137471283-6776b06f-438c-4fdc-9ad8-f6fa663b8e09.png)
_Current Presentation_

jniles committed:
feat(reports): purchase price report

This commit adds the purchase price report that shows the evolution of
our stock entry prices over time.  It also has a comparison field for
the current inventory price, and shows any negative margins in red.

Close #5519.

jmcameron commented on client/src/i18n/en/report.json:
```diff
@@ -157,6 +157,10 @@
       "TITLE": "Purchase Order Analysis"
     },
+    "PURCHASE_PRICES": {
+      "DESCRIPTION": "This report shows the purchase value of a medicine over time, eachtime it was entered into stock.  It also shows the margin over the current inventory price.",
```
Typo:   eachtime -> each time

jmcameron commented on client/src/modules/reports/reports.routes.js:
```diff
@@ -8,52 +8,53 @@ angular.module('bhima.routes')
       'account_report_multiple',
```
Thank you for sorting these!

jmcameron commented on server/models/bhima.sql:
```diff
@@ -434,7 +435,7 @@ INSERT INTO `indicator_status`(`id`, `text`,`translate_key`)VALUES
   (1, 'hospitalization', 'DASHBOARD.HOSPITALIZATION'),
   (2, 'staff', 'DASHBOARD.STAFF'),
-  (3, 'fianance', 'DASHBOARD.FINANCE');
+  (3, 'finance', 'DASHBOARD.FINANCE');
```
Good catch!

jmcameron replied:
However, I think you need to fix the same item in the migrate.sql file.

jmcameron reviewed (commented):
I took a casual look at the code and it seemed fine.
I ran test with the IMCK db.   Here are few comments about the report output:

- It is not clear what the $0.16 is.  Either label it or delete it.
- What does the pink line mean (are we getting paid to take drugs)?
- Rather than having the Margin and percent margin in one cell, I suggest breaking them into two columns.  I think it will make it cleaner and also make the Excel version more useful.
- Since the report is only for one inventory item, I suggest putting the inventory code/text as a subtitle (under "PURCHASE PRICE ANALYSIS" in a larger font and not have it as part of the table.
- I don't see min/max prices (as described in the PR)
- We recently increased the number of digits after the decimal for unit prices (due to a Vanga request, as I recall).   I suggest changing it from 2 to 4 (or making it an option).
- The report currently covers all time.   Does it make sense to allow selection of a range of dates?
- Should this be specific to a Depot?

![image](https://user-images.githubusercontent.com/62145/128830559-ce2df596-1569-49ea-bd06-66aebf90ef65.png)

jniles committed:
chore: address @jmcameron's review

This commit addresses all the critiques in @jmcameron's review.  The
min/max prices have been added in, the percentage moved to a separate
column, and the number of digits in the currency renderer increased.

jmcameron reviewed (approved):
Looks great.
One MINOR cosmetic thing (ignore if you wish):
![image](https://user-images.githubusercontent.com/62145/128849267-002de45d-f44d-4da2-895d-6c8484b5d6f2.png)

I don't think the "Sale Price" needs to be underlined.   However, adding a little CSS margin above it would make it a little easier to read.

mbayopanda commented on client/src/i18n/fr/report.json:
```diff
@@ -162,6 +162,10 @@
+    "PURCHASE_PRICES": {
+      "DESCRIPTION": "Ce rapport indique la valeur d'achat d'un médicament au fil du temps, à chaque entrée en stock. Il montre également la marge par rapport au prix d'inventaire actuel.",
+      "TITLE": "Analyse du prix d'achat"
+    },
```
👍🏽

mbayopanda commented on server/controllers/stock/reports/purchase_prices/index.js:
```diff
@@ -0,0 +1,89 @@
+      (inv.price - sm.unit_cost) AS margin,
+      ((inv.price - sm.unit_cost) / sm.unit_cost) AS percent_margin,
+      ((inv.price - sm.unit_cost) < 0) AS negative,
+      SUM(sm.unit_cost * sm.quantity) OVER (ORDER BY sm.date) AS running_total,
```
:-) window function

mbayopanda reviewed (commented):
The code is looking good to me, but I cannot understand where this value is coming from 

![image](https://user-images.githubusercontent.com/5445251/129025633-3274fdac-5b77-438c-add1-f510c7718859.png)

jmcameron reviewed (commented):
I tested this and the UI and result table seems to work okay, 

But I have a couple of questions to improve my understanding of what you are trying to display in this table.

- "Sales price" == inventory price, right?
- I think the margin should be per unit (not based on the total sale)
- I suggest moving the Total to the last column and and putting the margin/percent margin just after the unit price.  Also relabel "Margin" to "Unit Margin".
- What is **margin**?  When I hear the word margin, I think of what is received that is MORE than what it costs us.   So that would be:  Margin = Inventory Price - unit_price, right?  Doesn't this assume all stock sales will receive full inventory price and ignore discounts, patient group subsidies, etc, etc?
- I also wondered if the percentage margin be based on the inventory price.   Currently we are doing:  

      Percent Margin = 100% * (inventory_price - unit_price)/unit_price

   It seems to me it should be:

      Percent Margin = 100% * (inventory_price - unit_price) / inventory_price

   Since the unit price varies significantly, it distorts the margins/percent margins when it is small.   Using the inventory price seems more stable.

- What value is there in showing the running value and sum at the bottom in the "Total" column?  
- Shouldn't the value of the stock be based on WAC?

mbayopanda reviewed (approved):
Calculation in the report are correct 👍🏽 

LGTM

jniles committed:
Merge branch 'master' into 5519-feat-add-report-on-purchase-prices

jniles committed:
chore: fix broken merge

jniles committed:
chore: remove margins

Removes all references to margins to just show previous purchase prices.

jmcameron reviewed (commented):
Looks good.    
A few of comments:
- Totals (sums) do not need 4 decimal places; 2 is plenty.   We do need more for the unit price as you have done.
- It would be helpful if the drop-down menu to select an inventory item would indicate items which are not in stock and have no purchase history.  Not sure if the list should be pruned (easy) or non-existing items greyed out (not so easy).
- It would be useful of the price report could be accessed through the action menus on the inventory registry and articles in stock registry.

jniles committed:
chore: add links to report to inventory

Improves currency rendering and adds a link to the inventory list.

jniles committed:
Merge remote-tracking branch 'upstream/master' into 5519-feat-add-report-on-purchase-prices

jmcameron commented on server/controllers/stock/reports/purchase_prices/report.handlebars:
```diff
@@ -0,0 +1,92 @@
+        <table class="table table-condensed table-report table-bordered">
+          <tbody>
+            <tr>
+              <th>{{translate 'FORM.LABELS.MAX_PRICE'}}</th>
```
I suggest changing max/min price to:  Max unit price / Min unit price (to avoid confusion)

jniles replied:
Good idea.  I'll do that and release.

jmcameron reviewed (approved):
It looks great.  Very useful report.
One minor suggestion: at the end of the form, to avoid confusion, change
- Max price => Max unit price
- Min price => Min unit price

This PR is ready to be released.  If you want to make the change above, fine.  Otherwise I can fold it into the next PR I work on.

jniles committed:
Merge remote-tracking branch 'upstream/master' into 5519-feat-add-report-on-purchase-prices
